src/dag-node-factory.js

```diff
@@ -0,0 +1,139 @@
+'use strict'
+
+const stable = require('stable')
+const sort = stable.inplace
+const mh = require('multihashes')
+const util = require('./util')
+const DAGLink = require('./dag-link')
+
+class DAGNodeFactory {
+  static create (data, dagLinks, hashAlg) {
+    if (!dagLinks) {
+      dagLinks = []
+    }
+
+    if (!hashAlg) {
+      hashAlg = 'sha2-256'
+    }
+
+    const links = dagLinks.map((l) => {
+      if (!l.constructor && l.constructor.name !== 'DAGLink') {
+        return l
+      } 
+
+      // haadcode: are the .name vs .Name for backwards compatibility?
+      const link = new DAGLink(l.name || l.Name,
+                               l.size || l.Size,
+                               l.hash || l.Hash || l.multihash)
+      return link
+    })
+
+    // haadcode: would be better to have: links = sort(links, util.linkSort)
+    sort(links, util.linkSort)
+
+    return new Promise((resolve, reject) => {
+      util.serialize({ data: data, links: links}, (err, serialized) => {
+        if (err) {
+          reject(err)
+        }
+        util.hash(hashAlg, serialized, (err, multihash) => {
+          if (err) {
+            reject(err)
+          }
+          const dagNode = new DAGNode(data, links, serialized, multihash)
+          resolve(dagNode)
+        })
+      })
+
+      /* 
+        Alternatively, if util functions were promisified, 
+        we could write this much shorter as:
+        
+        return util.serialize({ data: data, links: links})
+          .then((serialized) => util.hash(hashAlg, serialized))
+          .then((multihash) => new DAGNode(data, links, serialized, multihash))
+      */
+    })
```

Comment:
I don't believe we are going to change this internal component to a Promised based one now. As we discussed in the past and again at the last gathering, so far the guideline is `callbacks at the internals, promises+callbacks at the top level`. 

This will only change if there is a huge discussion and decision here: https://github.com/ipfs/js-ipfs/issues/557